lib/util/quarantine/oaep.js

```diff
@@ -0,0 +1,203 @@
+// OAEP/MGF1 processes obviate the advantages of SHA2.
+//
+// because the use of SHA256 is so obscure, node's native crypto module (which
+// is a binding interface against an embedded OpenSSL) does not allow us to select
+// it as our padding hashing strategy.
+//
+// the openssl CLI tool will do it for you (but only if you use pkeyutl):
+// -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha256 -pkeyopt rsa_mgf1_md:sha256
+// so one method would be to shell out to it, since we have it everywhere (taking
+// care not to land the privkey on the disk). but macOS ships libreSSL instead
+// (which for the record i approve of), whose openssl CLI tool does not implement
+// these keyopts because they are not documented officially by OpenSSL. so to use
+// this approach would eliminate macOS as a native dev environment.
+//
+// the ursa library (which is also backed by OpenSSL) also does not allow the SHA256
+// hashing strategy we require.
+//
+// node-forge does, but having reviewed it i do not believe we can trust it. they've
+// done their best but they rely on third party code and at the end of the day they're
+// working in javascript. it is extraordinarily difficult to look at a piece of JS
+// and assert with absolute confidence that it never leaks via instruction pointers,
+// or branch indicators, or dependent array indices, or dynamic allocation, or that
+// it actually runs in constant time. it does not appear that any serious security
+// researcher has done an analysis on node-forge.
+//
+// that leaves us with three options, none of which are palatable but all of which
+// are more defensible than the above:
+//
+// 1. build a very thin binding layer to OpenSSL which sets the options we would need:
+//    EVP_PKEY_CTX_set_rsa_oaep_md(ctx, EVP_sha256());
+//    EVP_PKEY_CTX_set_rsa_mgf1_md(ctx, EVP_sha256());
+// 2. use node-forge, but sandbox it HEAVILY in a separate process and do our
+//    absolute best to pave over any timing/state uncertainty.
+// 3. implement our own OAEP unpadding routine.
+//
+// option 1 is very difficult, option 2 is a lot of work for very little cryptographic
+// gain. option 3 i would normally frown upon.
+//
+// !! !! !! !!! DO NOT IMPLEMENT YOUR OWN CRYPTOGRAPHIC PRIMITIVES !!! !! !! !!
+// !! !! !! !!! DO NOT IMPLEMENT YOUR OWN CRYPTOGRAPHIC PRIMITIVES !!! !! !! !!
+// !! !! !! !!! DO NOT IMPLEMENT YOUR OWN CRYPTOGRAPHIC PRIMITIVES !!! !! !! !!
+//
+// and yet, here we are, implementing our own cryptographic primitives. ultimately,
+// the reason this route was chosen was because we can make certain assumptions
+// in our particular use case that radically simplify the algorithm as typically
+// implemented:
+//
+// 1. the digest length will always be 256 bits. this fact is why we are on this
+//    whole stupid journey to begin with.
+// 2. the final cleartext length will always be 32 bytes. usually, once the final
+//    db is obtained, one must hunt for The One (the first 0x01 byte) starting from
```

Comment:
Pardon my ignorance -- `db` here and in `unpadPkcs1OaepMgf1Sha256` means "data buffer"?

Reply:
"data block", technically.